pkg/agent/containerd/watcher.go

```diff
@@ -0,0 +1,245 @@
+
+	return true
+}
+
+func (w *Watcher) processImageEvent(ctx context.Context, event fsnotify.Event, cfg *config.Node, client *containerd.Client, imageClient runtimeapi.ImageServiceClient) error {
+	defer w.workqueue.Done(event)
+
+	switch event.Op {
+	case fsnotify.Write:
+		newStateFile, err := os.Stat(event.Name)
+		if err != nil {
+			logrus.Errorf("Error encountered while getting file %s info for event write: %s", event.Name, err.Error())
+			break
+		}
+
+		// we do not want to handle directorys, only files
+		if newStateFile.IsDir() {
+			w.workqueue.Done(event)
+			break
+		}
+
+		lastStateFile := w.filesMap[event.Name]
+		w.filesMap[event.Name] = newStateFile
+
+		if (newStateFile.Size() != lastStateFile.Size()) && newStateFile.ModTime().After(lastStateFile.ModTime()) {
+			logrus.Infof("File met the requirements for import to containerd image store: %s", event.Name)
+			start := time.Now()
+			if err := preloadFile(ctx, cfg, client, imageClient, event.Name); err != nil {
+				logrus.Errorf("Error encountered while importing %s: %v", event.Name, err)
+				break
+			}
+			logrus.Infof("Imported images from %s in %s", event.Name, time.Since(start))
+		}
+	case fsnotify.Create:
+		info, err := os.Stat(event.Name)
+		if err != nil {
+			logrus.Errorf("Error encountered while getting file %s info for event Create: %v", event.Name, err)
+			break
+		}
+
+		if info.IsDir() {
+			w.workqueue.Done(event)
+			break
+		}
+
+		w.filesMap[event.Name] = info
+		logrus.Infof("File added to watcher controller: %s", event.Name)
+
+		start := time.Now()
+		if err := preloadFile(ctx, cfg, client, imageClient, event.Name); err != nil {
+			logrus.Errorf("Error encountered while importing %s: %v", event.Name, err)
+			break
+		}
+		logrus.Infof("Imported images from %s in %s", event.Name, time.Since(start))
+	case fsnotify.Rename:
+		// clear label because when a new file will be created because of the rename
+		// the create event will label the images with the name of the new file
+		err := clearLabelFromAutoImport(ctx, client, event.Name)
+		if err != nil {
+			logrus.Errorf("Failed to clear auto import label for %s: %v", event.Name, err)
+		}
+		delete(w.filesMap, event.Name)
+		logrus.Infof("Removed file from the watcher controller: %s", event.Name)
+	case fsnotify.Remove:
+		// Remove the auto import label from the image
+		err := clearLabelFromAutoImport(ctx, client, event.Name)
+		if err != nil {
+			logrus.Errorf("Failed to clear auto import label for %s: %v", event.Name, err)
+		}
+		delete(w.filesMap, event.Name)
+		logrus.Infof("Removed file from the watcher controller: %s", event.Name)
+	}
+
+	return nil
+
+}
+
+func watchImages(ctx context.Context, cfg *config.Node) {
+	w, err := CreateWatcher()
+	if err != nil {
+		logrus.Panicf("Failed to create watcher: %v", err)
+	}
+
+	logrus.Info("Watcher created")
+
+	defer w.watcher.Close()
+
+	if err := w.HandleWatch(cfg.Images); err != nil {
+		logrus.Panicf("Failed to watch %s: %v", cfg.Images, err)
```

Comment:
again, should probably just error and not watch instead of panicing in case fsnotify is not supported.